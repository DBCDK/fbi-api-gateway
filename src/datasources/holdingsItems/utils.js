// Helper function to construct the URL dynamically
export function buildPath(
  url,
  { agencyId, bibliographicRecordId, trackingId, itemId }
) {
  const basePath = [url, agencyId, bibliographicRecordId, itemId]
    .filter(Boolean)
    .join("/");
  const query = trackingId
    ? `?trackingId=${encodeURIComponent(trackingId)}`
    : "";
  return `${basePath}${query}`;
}
