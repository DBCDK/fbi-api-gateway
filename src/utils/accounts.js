/**
 * @file This file handles util function to select single or multiple CULR accounts
 *
 */

import { isValidCpr } from "./cpr";

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} Object or null
 */
export async function getAccount(accessToken, context, props) {
  return (await getAccounts(accessToken, context, props))?.[0] || null;
}

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} (Array)
 */
export async function getAccounts(accessToken, context, props) {
  // userInfo
  const userinfo = await context.datasources.getLoader("userinfo").load({
    accessToken,
  });

  const user = userinfo?.attributes;

  if (!user?.userId) {
    return [];
  }

  // select dataloader
  let dataloader = isValidCpr(user.userId)
    ? "culrGetAccountsByGlobalId"
    : "culrGetAccountsByLocalId";

  // Retrieve user culr account
  const response = await context.datasources.getLoader(dataloader).load({
    userId: user.userId,
    agencyId: user.loggedInAgencyId,
  });

  return filterAccountsByProps(response.accounts, props);
}

/**
 *
 * @param {Array} accounts
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAccountsByProps(accounts = [], props = {}) {
  if (props.agency) {
    accounts = accounts?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    accounts = accounts?.filter(({ userIdValue }) => userIdValue === props.id);
  }
  if (props.type) {
    accounts = accounts?.filter(({ userIdType }) => userIdType === props.type);
  }
  return accounts;
}

/**
 *
 * Function to select a specific agency from a userinfo agencies list
 *
 * @param {Array} agencies
 * @param {object} props - Filter agencies by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAgenciesByProps(agencies = [], props = {}) {
  if (props.agency) {
    agencies = agencies?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    agencies = agencies?.filter(({ userId }) => userId === props.id);
  }
  if (props.type) {
    agencies = agencies?.filter(({ userIdType }) => userIdType === props.type);
  }
  return agencies;
}
