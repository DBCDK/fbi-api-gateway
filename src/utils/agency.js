/**
 * Function to check if an agencyId is a FFU library
 *
 * @param {string} branchId
 *
 * @returns {boolean}
 */

export function _isFFUAgency(branchId) {
  const LENGTH = 6;
  const list = ["4", "6", "8", "9"];
  return branchId.length === LENGTH && list.includes(branchId.charAt(0));
}

/**
 * Function to check if an agencyId is a FFU library
 *
 * @param {string} branchId
 * @param {object} context
 *
 * @returns {boolean}
 */

export async function isFFUAgency(branchId, context) {
  if (!context) {
    // If no context given/available, fallback to old/static check.
    return _isFFUAgency(branchId);
  }

  const loader = context?.getLoader || context?.datasources?.getLoader;

  const result = (await loader("library").load({ branchId })).result?.[0];

  // toUpperCase needed for mocked agencies (jest testing)
  return !!(result?.agencyType?.toUpperCase() === "FORSKNINGSBIBLIOTEK");
}

/**
 *
 * @param {string} branchId
 * @param {object} context
 *
 * @returns {boolean}
 */

export async function isFolkAgency(branchId, context) {
  if (!context) {
    // If no context given/available, fallback to old/static check.
    return branchId.startsWith("7");
  }

  const loader = context?.getLoader || context?.datasources?.getLoader;

  const result = (await loader("library").load({ branchId })).result?.[0];

  // toUpperCase needed for mocked agencies (jest testing)
  return !!(result?.agencyType?.toUpperCase() === "FOLKEBIBLIOTEK");
}

/**
 * Function to check if an agency sync data with culr
 *
 * @param {string} branchId
 * @returns {boolean}
 */
export async function hasCulrDataSync(branchId, context) {
  /**
   * Odense Katedralskole, 872960,
   * Roskilde Gymnasium, 872600
   * Sorø Akademis Skole, 861640
   * Slagelse Gymnasium, biblioteket 872320
   * Greve Gymnasium, Biblioteket 874260
   * Stenhus Gymnasium, 875140
   * Sct. Knuds Gymnasium, 871890
   */

  const whitelist = [
    // Gymnasier
    "872960",
    "872600",
    "861640",
    "872320",
    "874260",
    "875140",
    "871890",

    // Login with mitId, when no library accounts
    "190101",
  ];

  if (whitelist.includes(branchId)) {
    return true;
  }

  return !!(await isFolkAgency(branchId, context));
}

/**
 * Function to check if an branch acts as independent agency
 *
 * @param {string} branchId // agencyId or branchId
 * @returns {boolean}
 */
export function branchIsIndependent(branchId) {
  const whitelist = [
    // Agencies
    "876040", // Nordjyske Gymnasiebiblioteker agency

    // Branches
    "872050", // Støvring Gymnasium, Biblioteket
    "872060", // Hjørring Gymnasium og HF-kursus, Biblioteket
    "872080", // Hasseris Gymnasium, Biblioteket
    "872090", // Brønderslev Gymnasium og HF-Kursus, Biblioteket
    "872100", // Vesthimmerlands Gymnasium & HF, Biblioteket
    "872140", // Aalborghus Gymnasium, Biblioteket
    "872340", // Frederikshavn Gymnasium, Gymnasiebiblioteket
    "872520", // Aalborg Katedralskole, Biblioteket
    "873310", // Nørresundby Gymnasium og HF, Biblioteket
    "874100", // Dronninglund Gymnasium, Biblioteket
  ];

  return whitelist.includes(branchId);
}

export async function getUserFromAllUserStatusData(props, context) {
  const agencies = props?.agencies || context?.user?.agencies;

  const all = await Promise.allSettled(
    agencies.map(async ({ agencyId, userId, userIdType }) => ({
      agencyId,
      userIdType,
      ...(await context.datasources.getLoader("user").load({
        userId,
        agencyId,
      })),
    }))
  );

  const users = [...all.map((u) => u?.value)];

  // remove undefined fields
  users.forEach((obj) =>
    Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key])
  );

  // Split data into CPR/LOCAL
  const cpr = users.filter((u) => u.userIdType === "CPR");
  const locals = users.filter((u) => u.userIdType === "LOCAL");

  // Prioritize CPR over LOCAL data
  return Object.assign({}, ...locals, ...cpr);
}

/**
 *
 * Function to find agencyId from branchId
 *
 * @param {string} branchId
 * @param {object} context
 * @returns {string} agencyId
 */

export async function getAgencyIdByBranchId(branchId, context) {
  const loader = context?.getLoader || context?.datasources?.getLoader;

  // get AgencyId from used branchId
  const result = (await loader("library").load({ branchId })).result?.[0];

  // return agencyId
  const agencyId = result?.agencyId;

  //  Return branchId instead of agencyId if branch act independently
  if (branchIsIndependent(branchId)) {
    return branchId;
  }

  return agencyId;
}
