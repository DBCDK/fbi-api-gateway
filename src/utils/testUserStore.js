/**
 * @file Redis Test User Store
 *
 * This file serves as a Redis store for managing test users. It is particularly useful
 * for creating test user scenarios where setting up a test environment is challenging, like:
 *
 * - nemlogin (where we don't have a test user)
 * - users on local libraries (where we for instance wants to be blocked)
 * - Merging different types of library accounts (FFU and Folkebibliotek) in Culr
 */

import { v5 as uuid } from "uuid";
import { set, get } from "../datasources/redis.datasource";
const TIME_TO_LIVE_SECONDS = 60 * 60 * 24 * 30;

/**
 * Generate a key from context, used to store data in Redis
 */
const getKey = (context) => `test_user_1:${context.testUser.key}`;

/**
 * Generate uuid deterministically based on a string
 */
function uuidFromString(str, context) {
  return (
    "test:" +
    uuid(getKey(context) + str, "1b671a64-40d5-491e-99b0-da01ff1f3341")
  );
}

/**
 * Store test user object in Redis
 */
export async function storeTestUser(object, context) {
  const getLoader = context?.datasources?.getLoader || context.getLoader;

  await set(getKey(context), TIME_TO_LIVE_SECONDS, object);

  // Clear redis entries
  await getLoader("userinfo").clearRedis({ accessToken: context.accessToken });
  if (object?.accounts.length > 0) {
    await Promise.all(
      object.accounts.map(async (account) => {
        await getLoader("borchk").clearRedis({
          userId: "123456",
          libraryCode: account.agency,
        });
      })
    );
  }
}

/**
 * Get test user object from Redis, and create a proper structured
 * test user object.
 */
export async function getTestUser(context) {
  let res = (await get(getKey(context)))?.val;

  // Get info for login agency
  const branch = (
    await (context?.datasources?.getLoader || context.getLoader)(
      "library"
    ).load({
      branchId: context.testUser.loginAgency,
      status: "ALLE",
    })
  ).result[0];

  if (!res?.accounts) {
    res = { accounts: [] };
  }

  const agencyId = branch?.agencyId || "190101";
  const cpr = branch?.agencyId ? null : "0101011234";
  const localId = "123456";

  if (!res?.accounts?.find((account) => account.agency === agencyId)) {
    res.accounts.push({ agency: agencyId, cpr });
    await storeTestUser(res, context);
  }

  res.accounts = res.accounts?.map((account) => ({
    ...account,
    localId,
    uniqueId: uuidFromString(account.cpr || account.agency || "", context),
  }));
  res.loginAgency = res?.accounts?.find(
    (account) => account.agency === agencyId
  );
  res.merged = res.accounts.filter(
    (account) => account.uniqueId === res.loginAgency.uniqueId
  );

  return res;
}

/**
 * Create accounts in Culr format
 */
export function accountsToCulr(accounts) {
  const res = [];
  accounts.forEach((account) => {
    res.push({
      agencyId: account.agency,
      userId: account.localId,
      userIdType: "LOCAL",
    });
    if (account.cpr) {
      res.push({
        agencyId: account.agency,
        userId: account.cpr,
        userIdType: "CPR",
      });
    }
  });
  return res;
}

export function parseTestToken(token) {
  const [testTokenType, accessToken] = token.split(":");
  const [_test, loginAgency, key] = testTokenType.split("_");
  return {
    accessToken,
    testUser: {
      loginAgency,
      key,
    },
  };
}
