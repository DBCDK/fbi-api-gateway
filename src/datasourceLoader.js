import DataLoader from "dataloader";

import { log } from "dbc-node-logger";
import { withRedis, clearRedis } from "./datasources/redis.datasource";
import { createFetchWithConcurrencyLimit } from "./utils/fetchWithLimit";
import { getFilesRecursive } from "./utils/utils";
import config from "./config";

/*
  This is how we time parallel requests to a single datasource.
  
  Request A
  ------------
          Request B
          ------------------
                                      Request C
                                      -----------
  Total AB
  --------------------------
  
  The totaltime for the requests should NOT be:
  Request A + Request B + Request C
  
  But instead:
  Total AB + Request C
*/

export const trackMe = () => {
  return {
    start: function (uuid) {
      this.uuid = uuid;
      return this;
    },
    createTracker: function (name) {
      // If this datasource has not been tracked yet, initialize its tracking object.
      if (!this.trackObject[name]) {
        this.trackObject[name] = {
          overlappingRequestCount: 0, // Number of ongoing requests for this datasource.
          count: 0, // Total number of elements retrieved from datasource
          started: 0, // Timestamp when the first concurrent request starts.
          time: 0, // Total time taken for all requests so far.
          sequentialTime: 0, // If requests were to be run sequentially, this is the total time
          batches: 0, // Dataloader batches multiple requests, into a single request when possible
        };
      }
    },

    // Function to indicate the beginning of a datasource call.
    begin: function (name, count) {
      this.createTracker(name);

      const beginTime = performance.now();

      const trackObject = this.trackObject[name];

      // Increment the count of ongoing requests.
      trackObject.overlappingRequestCount++;

      // Increment total number of elements retrieved from datasource
      trackObject.count += count;
      trackObject.batches++;

      // Function to indicate the end of an datasource call.
      return function end() {
        const endTime = performance.now();
        const duration = Math.round(endTime - beginTime);
        // Decrement the count of ongoing requests.
        trackObject.overlappingRequestCount--;

        // If all ongoing requests are complete, calculate the time taken.
        if (trackObject.overlappingRequestCount === 0) {
          trackObject.time += duration;
        }

        trackObject.sequentialTime += duration;
      };
    },

    cacheMiss: function (name, count) {
      this.createTracker(name);

      this.trackObject[name].cacheMiss =
        (this.trackObject[name].cacheMiss || 0) + count;
    },
    uuid: null,
    trackObject: {},
    getMetrics: function () {
      const res = {};

      Object.entries(this.trackObject).forEach(([key, val]) => {
        res[key] = {
          count: val.count,
          time: val.time,
          cacheMiss: val.cacheMiss,
          avg_ms: Math.round(val.sequentialTime / val.count),
        };
      });

      return res;
    },
  };
};

// Find all datasources in src/datasources
export const datasources = getFilesRecursive(`${__dirname}/datasources`)
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status, testLoad } = require(file.path);
    if (!load && !batchLoader) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    return {
      testLoad,
      load,
      batchLoader,
      name,
      options,
      statusChecker: status ? () => status(load) : null,
    };
  })
  .filter((func) => !!func);

const nameToDatasource = {};
datasources.forEach((ds) => (nameToDatasource[ds.name] = ds));

log.debug(
  `found ${datasources.length} datasources, ${datasources
    .map((datasource) => datasource.name)
    .join(", ")}`
);

/**
 * This will initialize a single dataloader.
 * It makes sure a context (with a concurrency limited fetcher), is injected
 * into the batchloader function.
 *
 * It sets up Redis for the dataloader
 *
 * It collects performance metrics for the datasource.
 */
function setupDataloader(
  { name, load, options, batchLoader, testLoad },
  context
) {
  const testModeActivated = context?.testUser && testLoad;
  const loadImpl = testModeActivated ? testLoad : load;

  let batchLoaderWithContext = batchLoader
    ? (keys) => batchLoader(keys, context)
    : (keys) => Promise.all(keys.map((key) => loadImpl(key, context)));

  if (!testModeActivated && options?.redis?.prefix && options?.redis?.ttl) {
    batchLoaderWithContext = withRedis(batchLoaderWithContext, {
      ...options.redis,
      ...context,
      datasourceName: name,
    });
  }

  async function batchLoaderWithTiming(keys) {
    const end = context?.track?.begin?.(name, keys.length);

    try {
      return await batchLoaderWithContext(keys);
    } catch (err) {
      // Wraps the load call in a try catch, in order to capture the
      // stack trace before dataloader is called. Otherwise we would
      // be limited to see the stack trace from the data sources, but
      // not the GraphQL resolvers.
      // https://brunoscheufler.com/blog/2021-07-25-improving-dataloader-stack-traces
      if (err instanceof Error && err.stack) {
        const currentStack = { stack: "" };
        Error.captureStackTrace(currentStack, batchLoaderWithTiming);
        err.stack += currentStack.stack.replace("Error", "\nBefore DataLoader");
      }
      throw err;
    } finally {
      end?.();
    }
  }

  const loader = new DataLoader(batchLoaderWithTiming, {
    // If key is an object, we stringify
    // to make it useful as a cache key
    cacheKeyFn: (key) => (typeof key === "object" ? JSON.stringify(key) : key),
    maxBatchSize: 100,
  });

  if (options?.redis?.prefix) {
    loader.clearRedis = (key) => clearRedis(options?.redis?.prefix, key);
  }

  return {
    loader,
  };
}

/**
 * Will instantiate dataloaders from datasources.
 * This should be done for every incoming GraphQL request
 */
export default function createDataLoaders(uuid, testUser, accessToken) {
  const result = {};
  const track = trackMe().start(uuid);

  result.trackingObject = track;
  const fetchWithConcurrencyLimit = createFetchWithConcurrencyLimit(
    config.fetchConcurrencyLimit
  );
  // Gets a loader by name.
  // A loader will be initialized first time it is called
  function getLoader(name) {
    if (!result[name]) {
      result[name] = setupDataloader(nameToDatasource[name], {
        getLoader,
        track,
        fetch: (url, options) => fetchWithConcurrencyLimit(url, options, name),
        trackingId: uuid,
        testUser,
        accessToken,
      })?.loader;
    }
    return result[name];
  }
  result.getLoader = getLoader;
  return result;
}

/**
 * For testing
 */
export function createMockedDataLoaders() {
  const mockedDatasources = {};
  getFilesRecursive("./src/datasources/mocked")
    .filter(
      (file) =>
        file.path.endsWith("datasource.mocked.js") && require(file.path).load
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.mocked.js", ""),
      load: require(file.path).load,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDatasources[loader.name] = loader;
    });

  mockedDatasources.getLoader = (name) => mockedDatasources[name];

  return mockedDatasources;
}

export function createTestUserDataLoaders() {
  const mockedDataLoaders = createMockedDataLoaders();
  getFilesRecursive("./src/datasources")
    .filter(
      (file) =>
        file.path.endsWith("datasource.js") && require(file.path).testLoad
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.js", ""),
      load: require(file.path).testLoad,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDataLoaders[loader.name] = loader;
    });

  return mockedDataLoaders;
}
