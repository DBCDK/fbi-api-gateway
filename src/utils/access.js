/**
 * @file access util functions
 */

import { filterDuplicateAgencies, resolveManifestation } from "./utils";

/**
 *
 * Check whether any of the users libraries has access to Infomedia.
 * The returned agencyId will be billed for infomedia access.
 *
 * First we check if the users login library has access to info media, if not,
 * we check if any other og the users associated accounts has access to infomedia,
 * if so, we return the first one found.
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
async function infomedia(context) {
  const user = context?.user;

  // get rights from idp
  const idpRights = await context.datasources.getLoader("idp").load("");

  // check if users loggedInAgency has infomedia access
  const loggedInAgencyId = user?.loggedInAgencyId;

  if (loggedInAgencyId && idpRights[loggedInAgencyId]) {
    return loggedInAgencyId;
  }

  /*
   * Alternativly check all users accounts
   */

  // filtered accounts
  const userInfoAccounts = filterDuplicateAgencies(user.agencies);

  const hasAccess = userInfoAccounts?.filter(
    ({ agencyId }) => idpRights[agencyId]
  );

  // check for infomedia access - if any of users agencies subscribes
  if (hasAccess.length > 0) {
    return hasAccess[0]?.agencyId;
  }

  return null;
}

/**
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
export async function getInfomediaAgencyId(context) {
  return await infomedia(context);
}

/**
 *
 * @param {object} context
 *
 * @returns {boolean}
 */
export async function hasInfomediaAccess(context) {
  return !!(await infomedia(context));
}

/**
 * Get type of access for manifestation with given pid.
 *  url access
 *  ill access
 *  infomedia access
 *  digital article service access
 *
 * @param pid
 * @param context
 * @returns {Promise<*>}
 */
export async function resolveAccess(manifestation, context) {
  // We parse the access structure from JED, and convert it
  // to the union type structure.
  // At some point we may choose to follow the structure of JED closely,
  // but it has to be coordinated with stakeholders

  let parent =
    typeof manifestation === "object"
      ? manifestation
      : await resolveManifestation({ pid: manifestation }, context);

  const res = [];

  // use linkchecker to check status of a single accessUrl
  // linkchecker does NOT handle proxy urls, so we skip ebookcentral and ebsco urls
  const linkStatus = async (url) => {
    if (url?.includes("ebookcentral") || url?.includes("ebscohost")) {
      return "OK";
    }

    const linkcheck = await context.datasources
      .getLoader("linkcheck")
      .load({ urls: [url] });
    const link = linkcheck.find((check) => check.url === url);

    return link?.status || "OK";
  };

  parent?.access?.accessUrls?.forEach((entry) => {
    const { proxyUrl, loginRequired } = getProxyUrl(
      entry.url || "",
      context?.user
    );

    res.push({
      __typename: "AccessUrl",
      origin: parseOnlineUrlToOrigin(entry.url),
      url: proxyUrl,
      loginRequired,
      note: entry.note,
      type: entry.type,
      status: linkStatus(proxyUrl) || "OK",
    });
  });

  if (parent?.access?.dbcWebArchive) {
    const archives = await context.datasources
      .getLoader("moreinfoWebarchive")
      .load(parent.pid);

    archives.forEach((archive) => {
      if (archive.url) {
        res.push({
          __typename: "AccessUrl",
          url: archive.url,
          origin: parseOnlineUrlToOrigin(archive.url),
          loginRequired: false,
          type: "RESOURCE",
          status: linkStatus(archive.url) || "OK",
        });
      }
    });
  }

  parent?.access?.ereol?.forEach((entry) => {
    res.push({
      __typename: "Ereol",
      origin: entry.origin,
      url: entry.url,
      canAlwaysBeLoaned: entry.canAlwaysBeLoaned,
      note: entry.note,
    });
  });

  if (parent?.access?.infomediaService?.id) {
    if (
      !["politiken", "jyllands-posten"].some((publication) =>
        parent?.hostPublication?.title?.toLowerCase()?.includes(publication)
      )
    ) {
      res.push({
        __typename: "InfomediaService",
        id: parent.access.infomediaService.id,
      });
    }

    // If this is a omitted online access for politiken or jyllands-posten.
    else {
      const hasPhysical = parent?.accessTypes?.some(
        (t) => (t?.code ?? "").toUpperCase() === "PHYSICAL"
      );

      if (hasPhysical) {
        parent = {
          ...parent,
          access: {
            ...(parent.access ?? {}),
            interLibraryLoanIsPossible: true,
          },
        };
      }
    }
  }

  // Get the issn for this article or periodica
  const issn =
    parent?.access?.digitalArticleService?.issn ||
    parent?.identifiers?.find?.((id) => id.type === "ISSN")?.value;

  if (issn) {
    const journals = await context.datasources
      .getLoader("statsbiblioteketJournals")
      .load("");
    const articleissn = issn.replace(/[^a-z\d]/gi, "");
    const hasJournal = journals && journals[articleissn];
    if (hasJournal) {
      res.push({
        __typename: "DigitalArticleService",
        issn: articleissn,
      });
    }
  }

  if (parent?.access?.interLibraryLoanIsPossible) {
    //if collectionidentifiers includes 870970-accessnew it may be loaned
    const isNewMaterial =
      parent?.collectionIdentifiers?.includes("870970-accessnew");

    const forLoan = checkInterLibraryLoan(parent, context);
    res.push({
      __typename: "InterLibraryLoan",
      loanIsPossible: forLoan,
      accessNew: isNewMaterial,
    });
  }

  // Return array containing all types of access
  return _sortOnlineAccess(res);
}

/**
 * municipality number is the second|third|fourth digit in agencyId
 * @param agencyId
 * @returns {string | undefined}
 */
function parseForMunicipalityNumber(agencyId) {
  return agencyId?.substring(1, 4);
}

/**
 * This one is for ebook.plus - we need to go via a proxy url (if user is logged in)
 * @param url
 * @param user
 * @returns {*}
 */
export function getProxyUrl(url, user) {
  const municipality =
    user?.municipality ||
    parseForMunicipalityNumber(user?.municipalityAgencyId);

  // check if we should proxy this url - for now it is ebookcentral and ebscohost
  const proxyMe =
    url.indexOf("ebookcentral") !== -1 || url.indexOf("ebscohost") !== -1;
  if (proxyMe) {
    // check if user is logged in
    if (user?.userId) {
      const realUrl = `https://bib${municipality}.bibbaser.dk/login?url=${url}`;
      return { proxyUrl: realUrl, loginRequired: proxyMe };
    }
  }

  return { proxyUrl: url, loginRequired: proxyMe };
}

/**
 * Some sources are not for loan (fagbibliografier) - for now a statis list:
 * This method is only called if interLibraryLoan is true - We know that for a fact
 * 159080-fagbib - Besættelsesbibliografien
 * 159081-fagbib - Bibliografi over Dansk Kunst
 * 159082-fagbib - Dansk Historisk Bibliografi
 * 159083-fagbib - Dansk Musiklitterær Bibliografi
 * 159084-fagbib - Kongelige Teater programartikler
 * 159085-fagbib - Dania Polyglotta
 * 159086-fagbib - Sportline
 * @param parent
 * @param context
 */
function checkInterLibraryLoan(parent, context) {
  // This is a blacklist of bases NOT to be loaned
  const notForLoanList = [
    "159080-fagbib",
    "159081-fagbib",
    "159082-fagbib",
    "159083-fagbib",
    "159084-fagbib",
    "159085-fagbib",
    "159086-fagbib",
  ];
  // we check on objectId - first part is the source
  const source = parent?.objectId?.split(":")[0];
  return !notForLoanList.includes(source);
}

/**
 * Get domain from url
 * @param url
 * @return String
 *  A parsed string eg. "DBC Webarkiv" or name of host eg. "infolink2003.elbo.dk"
 *
 * */
export function parseOnlineUrlToOrigin(url) {
  try {
    const parsedUrl = new URL(url);
    if (parsedUrl["host"] === "moreinfo.addi.dk") {
      return "DBC Webarkiv";
    } else {
      return (parsedUrl["host"] && parsedUrl["host"]) || "";
    }
  } catch (e) {
    return "";
  }
}

/**
 * Handle special cases - for now we want filmstriben/fjernleje BEFORE filmstriben/biblioteket AND
 * we want dansk film institut (dfi) to come last - it is not a 'real' url but a link to a description
 * @param onlineAccess
 * @return {*}
 * @private
 */
function _sortOnlineAccess(onlineAccess) {
  const specialSort = (a, b) => {
    // fjernleje should be on top
    if (b.url && b.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return 1;
    } else if (a.url && a.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return -1;
      // dfi is not a 'real' online url - sort low
    } else if (b.url && b.url.indexOf("dfi.dk") !== -1) {
      return -1;
    } else if (a.url && a.url.indexOf("dfi.dk") !== -1) {
      return 1;
    }
    return 0;
  };

  return onlineAccess.sort(specialSort);
}
