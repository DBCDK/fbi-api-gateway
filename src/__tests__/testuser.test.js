import { createTestUserDataLoaders } from "../datasourceLoader";
import { parseTestToken } from "../utils/testUserStore";
import { performTestQuery } from "../utils/utils";

const redisObjects = {};
jest.mock("../datasources/redis.datasource", () => ({
  get: (key) => redisObjects[key],
  set: (key, _ttl, val) => (redisObjects[key] = { val }),
}));
const testToken = "test_nemlogin_kfu:d66828f989e1184g0d51g1dcdf9d9985ecdcb6ed";
const testUserQuery = `query TestUser {
  test {
    user {
      accounts {
        agency {
          agencyName
          agencyId
        }
        cpr
        debt
        uniqueId
        blocked
        isMunicipality
      }
    }
  }
}
`;

const testUserMutate = `mutation MutateTestUser($input: TestUserInput) {
  test {
    user(
      input: $input
    )
  }
}
`;

describe("Testuser", () => {
  test("parser can parse test token", async () => {
    expect(parseTestToken(testToken)).toMatchSnapshot();
  });
  test("test user can be fetched", async () => {
    const result = await performTestQuery({
      query: testUserQuery,
      variables: {},
      context: {
        ...parseTestToken(testToken),
        datasources: createTestUserDataLoaders(),
      },
    });
    expect(result).toMatchSnapshot();
  });
  test("test user is null without test token", async () => {
    const result = await performTestQuery({
      query: testUserQuery,
      variables: {},
      context: {
        accessToken: "d66828f989e118430d5151dcdf9d9985ecdcb6ed",
        datasources: createTestUserDataLoaders(),
      },
    });
    expect(result).toMatchSnapshot();
  });
  test("can store test user", async () => {
    await performTestQuery({
      query: testUserMutate,
      variables: {
        input: {
          accounts: [
            {
              isMunicipality: true,
              agency: "716300",
              debt: "100",
              cpr: "0101011234",
            },
            { agency: "715100" },
          ],
        },
      },
      context: {
        ...parseTestToken(testToken),
        datasources: createTestUserDataLoaders(),
      },
    });

    const result = await performTestQuery({
      query: testUserQuery,
      variables: {},
      context: {
        ...parseTestToken(testToken),
        datasources: createTestUserDataLoaders(),
      },
    });
    expect(result).toMatchSnapshot();
  });

  test("Error when storing test user without test token", async () => {
    const result = await performTestQuery({
      query: testUserMutate,
      variables: {
        input: {
          accounts: [
            {
              agency: "716300",
            },
            { agency: "715100" },
          ],
        },
      },
      context: {
        accessToken: "d66828f989e118430d5151dcdf9d9985ecdcb6ed",
        datasources: createTestUserDataLoaders(),
      },
    });
    expect(result).toMatchSnapshot();
  });
});
