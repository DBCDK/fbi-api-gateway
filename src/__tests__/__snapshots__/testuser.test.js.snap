// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testuser Error when storing test user without test token 1`] = `
Object {
  "data": Object {
    "test": Object {
      "user": "Not a valid test token",
    },
  },
}
`;

exports[`Testuser can store test user 1`] = `
Object {
  "data": Object {
    "test": Object {
      "user": Object {
        "accounts": Array [
          Object {
            "agency": Object {
              "agencyId": "716300",
              "agencyName": "Herlev Bibliotek",
            },
            "blocked": false,
            "cpr": "0101011234",
            "debt": "100",
            "isMunicipality": true,
            "uniqueId": "test:469b0831-1eef-5aca-bbe6-f6519fd922c9",
          },
          Object {
            "agency": Object {
              "agencyId": "715100",
              "agencyName": "Ballerup Bibliotekerne",
            },
            "blocked": false,
            "cpr": null,
            "debt": null,
            "isMunicipality": false,
            "uniqueId": "test:ce0da062-c974-5150-9749-252adcd92db7",
          },
          Object {
            "agency": Object {
              "agencyId": "190101",
              "agencyName": "nemlogin",
            },
            "blocked": false,
            "cpr": "0101011234",
            "debt": null,
            "isMunicipality": false,
            "uniqueId": "test:469b0831-1eef-5aca-bbe6-f6519fd922c9",
          },
        ],
      },
    },
  },
}
`;

exports[`Testuser parser can parse test token 1`] = `
Object {
  "accessToken": "d66828f989e1184g0d51g1dcdf9d9985ecdcb6ed",
  "testUser": Object {
    "key": "kfu",
    "loginAgency": "nemlogin",
  },
}
`;

exports[`Testuser test user can be fetched 1`] = `
Object {
  "data": Object {
    "test": Object {
      "user": Object {
        "accounts": Array [
          Object {
            "agency": Object {
              "agencyId": "190101",
              "agencyName": "nemlogin",
            },
            "blocked": false,
            "cpr": "0101011234",
            "debt": null,
            "isMunicipality": false,
            "uniqueId": "test:469b0831-1eef-5aca-bbe6-f6519fd922c9",
          },
        ],
      },
    },
  },
}
`;

exports[`Testuser test user is null without test token 1`] = `
Object {
  "data": Object {
    "test": Object {
      "user": null,
    },
  },
}
`;
